File: games/story/adapters/go/aaa_systems.go
package bizxgame

import "math/rand"

type WorldCell struct { ID string; Seed int64; Biome string; Population int; Weather string; Loaded bool }
type NPC struct { ID, Faction string; Food, Rest, Safety float64; State string }
type AAASystems struct { Tick int64; Cells map[string]WorldCell; NPCs map[string]*NPC; Prices map[string]float64; Events []string; RNG *rand.Rand }
func NewAAASystems(seed int64)*AAASystems{return &AAASystems{Cells:map[string]WorldCell{},NPCs:map[string]*NPC{},Prices:map[string]float64{},RNG:rand.New(rand.NewSource(seed))}}
func(s *AAASystems) AddCell(c WorldCell){s.Cells[c.ID]=c};func(s *AAASystems) AddNPC(n *NPC){s.NPCs[n.ID]=n}
func(s *AAASystems) TickOnce(){s.Tick++;for _,n:=range s.NPCs{n.Food=max(0,n.Food-.005);n.Rest=max(0,n.Rest-.003);if n.Safety<.25{n.State="flee"}else if n.Food<.25{n.State="seek_food"}else if n.Rest<.2{n.State="sleep"}else{n.State="work"}}}
func(s *AAASystems) Market(good string,supply,demand float64)float64{p:=max(.01,(demand+1)/(supply+1));s.Prices[good]=p;return p}
func(s *AAASystems) ResolveAbility(base,defense int)int{r:=s.RNG.Intn(20)+1;if r+5<defense{return 0};if r==20{return base*2};return base}
func max(a,b float64)float64{if a>b{return a};return b}

File: games/story/adapters/rust/aaa_systems.rs
#[derive(Clone,Debug)] pub struct WorldCell{pub id:String,pub seed:u64,pub biome:String,pub population:u32,pub weather:String,pub loaded:bool}
#[derive(Clone,Debug)] pub struct Npc{pub id:String,pub faction:String,pub food:f32,pub rest:f32,pub safety:f32,pub state:String}
pub struct AaaSystems{pub tick:u64,pub cells:Vec<WorldCell>,pub npcs:Vec<Npc>,pub prices:std::collections::HashMap<String,f32>}
impl AaaSystems{pub fn new()->Self{Self{tick:0,cells:vec![],npcs:vec![],prices:Default::default()}}
pub fn tick_once(&mut self){self.tick+=1;for n in &mut self.npcs{n.food=(n.food-.005).max(0.);n.rest=(n.rest-.003).max(0.);n.state=if n.safety<.25{"flee"}else if n.food<.25{"seek_food"}else if n.rest<.2{"sleep"}else{"work"}.into();}}
pub fn market(&mut self,good:&str,supply:f32,demand:f32)->f32{let p=((demand+1.)/(supply+1.)).max(.01);self.prices.insert(good.into(),p);p}
pub fn resolve_ability(&self,base:i32,defense:i32,roll:i32)->i32{if roll+5<defense{0}else if roll==20{base*2}else{base}}}
